tests/AsyncTest.php

```diff
@@ -44,4 +83,44 @@ public function testInstanceOfInterface() {
     $xhp = <async:test><b>BE BOLD</b></async:test>;
     $this->assertInstanceOf(XHPAwaitable::class, $xhp);
   }
+
+  public function parallelizationContainersProvider() {
+    return [
+      [<test:xfrag-wrap />],
+      [<test:async-xfrag-wrap />],
+    ];
+  }
+
+  /**
+   * @dataProvider parallelizationContainersProvider
+   */
+  public function testParallelization(:x:element $container) {
+    :async:par-test::$log = Vector { };
+
+    $a = <async:par-test label="a" />;
+    $b = <async:par-test label="b" />;
+    $c = <async:par-test label="c" />;
+
+    $container->replaceChildren([$b, $c]);
+
+    $tree = <async:test>{$a}{$container}</async:test>;
+    $this->assertSame('<div><div>a</div><div>b</div><div>c</div></div>', $tree->toString());
+
+    $log = :async:par-test::$log;
+    $by_node = Map { 'a' => Map { }, 'b' => Map { }, 'c' => Map { } };
+
+    foreach ($log as $idx => $data) {
+      list($label, $action) = $data;
+      $by_node[$label][$action] = $idx;
+    }
+
+    $max_start = max($by_node->map($x ==> $x['start']));
+    $min_mid = min($by_node->map($x ==> $x['mid']));
+    $max_mid = max($by_node->map($x ==> $x['mid']));
+    $min_finish = min($by_node->map($x ==> $x['finish']));
+
+    $this->assertGreaterThan($max_start, $min_mid, 'all should be started before any get continued');
+    $this->assertGreaterThan($max_mid, $min_finish, 'all should have reached stage two before any finish');
+    $this->assertGreaterThan($max_start, $min_finish, 'sanity check: all have started before any finish');
```

Comment:
This is actually the only one that matches HHVM's guarantees - the others match as long as HHVM keeps on eagerly executing until it sees a native waithandle, which is likely (though not certain) to continue.